File: src/lexer/lexer.zig
const std = @import("std");

pub const TokenType = enum {
    ILLEGAL,
    EOF,
    IDENT,
    INT,
    ASSIGN,
    PLUS,
    COMMA,
    SEMICOLON,
    LPAREN,
    RPAREN,
    LBRACE,
    RBRACE,
    FUNCTION,
    LET,
    MINUS,
    BANG,
    ASTERISK,
    SLASH,
    GT,
    LT,
    TRUE,
    FALSE,
    ELSE,
    IF,
    RETURN,
    EQ,
    NOT_EQ,
};

pub const Token = struct {
    const Self = @This();
    kind: TokenType,
    literal: []const u8,

    pub fn new(kind: TokenType, literal: []const u8) Self {
        return Self{
            .kind = kind,
            .literal = literal,
        };
    }
};

fn isLetter(ch: u8) bool {
    return std.ascii.isAlphabetic(ch) or ch == '_';
}

fn isDigit(ch: u8) bool {
    return std.ascii.isDigit(ch);
}

pub const Lexer = struct {
    const Self = @This();

    input: []const u8,
    position: usize = 0,
    readPosition: usize = 0,
    ch: u8 = 0,

    pub fn init(input: []const u8) Self {
        var lexer = Self{ .input = input };

        lexer.readChar();

        return lexer;
    }

    pub fn readChar(self: *Self) void {
        if (self.readPosition >= self.input.len) {
            self.ch = 0;
        } else {
            self.ch = self.input[self.readPosition];
        }
        self.position = self.readPosition;
        self.readPosition += 1;
    }

    pub fn nextToken(self: *Self) Token {
        self.skipWhitespace();
        const token: Token = switch (self.ch) {
            '=' => blk: {
                if (self.peekChar() == '=') {
                    self.readChar();
                    break :blk Token.new(.EQ, "==");
                } else {
                    break :blk Token.new(.ASSIGN, "=");
                }
            },
            ';' => Token.new(.SEMICOLON, ";"),
            '(' => Token.new(.LPAREN, "("),
            ')' => Token.new(.RPAREN, ")"),
            ',' => Token.new(.COMMA, ","),
            '+' => Token.new(.PLUS, "+"),
            '{' => Token.new(.LBRACE, "{"),
            '}' => Token.new(.RBRACE, "}"),
            '/' => Token.new(.SLASH, "/"),
            '!' => blk: {
                if (self.peekChar() == '=') {
                    self.readChar();
                    break :blk Token.new(.NOT_EQ, "!=");
                } else {
                    break :blk Token.new(.BANG, "!");
                }
            },
            '-' => Token.new(.MINUS, "-"),
            '>' => Token.new(.GT, ">"),
            '<' => Token.new(.LT, "<"),
            '*' => Token.new(.ASTERISK, "*"),
            'a'...'z', 'A'...'Z', '_' => {
                const ident = self.readIdentifier();
                if (self.lookupKeyword(ident)) |token| {
                    return token;
                }
                return Token.new(.IDENT, ident);
            },
            '0'...'9' => {
                const int = self.readDigit();
                return Token.new(.INT, int);
            },
            0 => Token.new(.EOF, ""),
            else => return Token.new(.ILLEGAL, ""),
        };

        self.readChar();
        return token;
    }

    fn lookupKeyword(self: *Self, ident: []const u8) ?Token {
        _ = self;
        const map = std.StaticStringMap(Token).initComptime(.{
            .{ "let", Token.new(.LET, "let") },
            .{ "fn", Token.new(.FUNCTION, "fn") },
            .{ "return", Token.new(.RETURN, "return") },
            .{ "if", Token.new(.IF, "if") },
            .{ "else", Token.new(.ELSE, "else") },
            .{ "true", Token.new(.TRUE, "true") },
            .{ "false", Token.new(.FALSE, "false") },
        });
        return map.get(ident);
    }

    fn peekChar(self: *Self) u8 {
        if (self.readPosition >= self.input.len) {
            return 0;
        } else {
            return self.input[self.readPosition];
        }
    }

    fn readIdentifier(self: *Self) []const u8 {
        const initialPosition = self.position;
        while (isLetter(self.ch)) {
            self.readChar();
        }
        return self.input[initialPosition..self.position];
    }

    fn readDigit(self: *Self) []const u8 {
        const initialPosition = self.position;
        while (isDigit(self.ch)) {
            self.readChar();
        }
        return self.input[initialPosition..self.position];
    }

    fn skipWhitespace(self: *Self) void {
        while (self.ch == '\r' or self.ch == ' ' or self.ch == '\n' or self.ch == '\t') {
            self.readChar();
        }
    }
};

test "Test all syntax" {
    const input =
        \\let five = 5;
        \\let ten = 10;
        \\
        \\let add = fn(x, y) {
        \\    x + y;
        \\};
        \\
        \\let result = add(five, ten);
        \\!-/*5;
        \\5 < 10 > 5;
        \\
        \\if (5 < 10) {
        \\   return true;
        \\} else {
        \\    return false;
        \\}
        \\10 == 10;
        \\10 != 9;
    ;
    const expectedTokens = [_]Token{
        Token.new(.LET, "let"),
        Token.new(.IDENT, "five"),
        Token.new(.ASSIGN, "="),
        Token.new(.INT, "5"),
        Token.new(.SEMICOLON, ";"),
        Token.new(.LET, "let"),
        Token.new(.IDENT, "ten"),
        Token.new(.ASSIGN, "="),
        Token.new(.INT, "10"),
        Token.new(.SEMICOLON, ";"),
        Token.new(.LET, "let"),
        Token.new(.IDENT, "add"),
        Token.new(.ASSIGN, "="),
        Token.new(.FUNCTION, "fn"),
        Token.new(.LPAREN, "("),
        Token.new(.IDENT, "x"),
        Token.new(.COMMA, ","),
        Token.new(.IDENT, "y"),
        Token.new(.RPAREN, ")"),
        Token.new(.LBRACE, "{"),
        Token.new(.IDENT, "x"),
        Token.new(.PLUS, "+"),
        Token.new(.IDENT, "y"),
        Token.new(.SEMICOLON, ";"),
        Token.new(.RBRACE, "}"),
        Token.new(.SEMICOLON, ";"),
        Token.new(.LET, "let"),
        Token.new(.IDENT, "result"),
        Token.new(.ASSIGN, "="),
        Token.new(.IDENT, "add"),
        Token.new(.LPAREN, "("),
        Token.new(.IDENT, "five"),
        Token.new(.COMMA, ","),
        Token.new(.IDENT, "ten"),
        Token.new(.RPAREN, ")"),
        Token.new(.SEMICOLON, ";"),
        Token.new(.BANG, "!"),
        Token.new(.MINUS, "-"),
        Token.new(.SLASH, "/"),
        Token.new(.ASTERISK, "*"),
        Token.new(.INT, "5"),
        Token.new(.SEMICOLON, ";"),
        Token.new(.INT, "5"),
        Token.new(.LT, "<"),
        Token.new(.INT, "10"),
        Token.new(.GT, ">"),
        Token.new(.INT, "5"),
        Token.new(.SEMICOLON, ";"),
        Token.new(.IF, "if"),
        Token.new(.LPAREN, "("),
        Token.new(.INT, "5"),
        Token.new(.LT, "<"),
        Token.new(.INT, "10"),
        Token.new(.RPAREN, ")"),
        Token.new(.LBRACE, "{"),
        Token.new(.RETURN, "return"),
        Token.new(.TRUE, "true"),
        Token.new(.SEMICOLON, ";"),
        Token.new(.RBRACE, "}"),
        Token.new(.ELSE, "else"),
        Token.new(.LBRACE, "{"),
        Token.new(.RETURN, "return"),
        Token.new(.FALSE, "false"),
        Token.new(.SEMICOLON, ";"),
        Token.new(.RBRACE, "}"),
        Token.new(.INT, "10"),
        Token.new(.EQ, "=="),
        Token.new(.INT, "10"),
        Token.new(.SEMICOLON, ";"),
        Token.new(.INT, "10"),
        Token.new(.NOT_EQ, "!="),
        Token.new(.INT, "9"),
        Token.new(.SEMICOLON, ";"),
        Token.new(.EOF, ""),
    };

    var lexer = Lexer.init(input);
    for (expectedTokens) |token| {
        const tokenFromLexer = lexer.nextToken();

        try std.testing.expectEqualDeep(token, tokenFromLexer);
    }
}

File: rust/src/lexer/lexer.rs
use std::str;

#[derive(Debug, PartialEq)]
pub enum Token {
    Illegal,
    Eof,
    Ident(String),
    Int(String),
    Assign,
    Plus,
    Comma,
    Semicolon,
    LParen,
    RParen,
    LBrace,
    RBrace,
    Function,
    Let,
    Minus,
    Bang,
    Asterisk,
    Slash,
    GT,
    LT,
    True,
    False,
    Else,
    If,
    Return,
    Eq,
    NotEq,
}

pub struct Lexer {
    pub input: Vec<u8>,
    pub position: usize,
    pub read_position: usize,
    pub ch: u8,
}

impl Lexer {
    fn init(input: String) -> Lexer {
        let mut lexer = Lexer {
            input: input.into_bytes(),
            position: 0,
            read_position: 0,
            ch: 0,
        };
        lexer.read_char();
        return lexer;
    }

    fn read_char(&mut self) {
        if self.read_position >= self.input.len() {
            self.ch = 0;
        } else {
            self.ch = *self.input.get(self.read_position).unwrap();
        }
        self.position = self.read_position;
        self.read_position += 1;
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_ascii_whitespace() {
            self.read_char();
        }
    }
    pub fn next_token(&mut self) -> Token {
        self.skip_whitespace();
        let token = match self.ch {
            b';' => Token::Semicolon,
            b'(' => Token::LParen,
            b')' => Token::RParen,
            b',' => Token::Comma,
            b'+' => Token::Plus,
            b'{' => Token::LBrace,
            b'}' => Token::RBrace,
            b'/' => Token::Slash,
            b'=' => {
                if self.peek_char() == b'=' {
                    self.read_char();
                    Token::Eq
                } else {
                    Token::Assign
                }
            }
            b'!' => {
                if self.peek_char() == b'=' {
                    self.read_char();
                    Token::NotEq
                } else {
                    Token::Bang
                }
            }
            b'*' => Token::Asterisk,
            b'-' => Token::Minus,
            b'>' => Token::GT,
            b'<' => Token::LT,
            b'a'..=b'z' | b'A'..=b'Z' | b'_' => {
                let ident = self.read_identifier();
                return match ident {
                    "let" => Token::Let,
                    "fn" => Token::Function,
                    "return" => Token::Return,
                    "if" => Token::If,
                    "else" => Token::Else,
                    "true" => Token::True,
                    "false" => Token::False,
                    _ => Token::Ident(ident.into()),
                };
            }
            b'0'..=b'9' => return Token::Int(self.read_digit().into()),
            0 => Token::Eof,
            _ => Token::Illegal,
        };

        self.read_char();
        return token;
    }

    fn peek_char(&self) -> u8 {
        if self.read_position >= self.input.len() {
            return 0;
        } else {
            return *self.input.get(self.read_position).unwrap();
        }
    }
    fn read_digit(&mut self) -> &str {
        let initial_position = self.position;
        while self.ch.is_ascii_digit() {
            self.read_char()
        }
        // cursed?
        return str::from_utf8(&self.input[initial_position..self.position]).unwrap();
    }

    fn read_identifier(&mut self) -> &str {
        let initial_position = self.position;
        while self.ch.is_ascii_alphabetic() || self.ch == b'_' {
            self.read_char()
        }
        // cursed?
        return str::from_utf8(&self.input[initial_position..self.position]).unwrap();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn init_works() {
        let input = "ok";
        let lexer = Lexer::init(input.into());
        assert_eq!(lexer.input, vec![b'o', b'k']);
        assert_eq!(lexer.position, 0);
        assert_eq!(lexer.read_position, 1);
        assert_eq!(lexer.ch, b'o');
    }

    #[test]
    fn test_all_syntax() {
        let input = r#"
            let five = 5;
            let ten = 10;

            let add = fn(x, y) {
                x + y;
            };

            let result = add(five, ten);
            !-/*5;
            5 < 10 > 5;

            if (5 < 10) {
                return true;
            } else {
                return false;
            }

            10 == 10;
            10 != 9;
        "#;
        let expected_tokens = vec![
            Token::Let,
            Token::Ident(String::from("five")),
            Token::Assign,
            Token::Int(String::from("5")),
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("ten")),
            Token::Assign,
            Token::Int(String::from("10")),
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("add")),
            Token::Assign,
            Token::Function,
            Token::LParen,
            Token::Ident(String::from("x")),
            Token::Comma,
            Token::Ident(String::from("y")),
            Token::RParen,
            Token::LBrace,
            Token::Ident(String::from("x")),
            Token::Plus,
            Token::Ident(String::from("y")),
            Token::Semicolon,
            Token::RBrace,
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("result")),
            Token::Assign,
            Token::Ident(String::from("add")),
            Token::LParen,
            Token::Ident(String::from("five")),
            Token::Comma,
            Token::Ident(String::from("ten")),
            Token::RParen,
            Token::Semicolon,
            Token::Bang,
            Token::Minus,
            Token::Slash,
            Token::Asterisk,
            Token::Int("5".into()),
            Token::Semicolon,
            Token::Int("5".into()),
            Token::LT,
            Token::Int("10".into()),
            Token::GT,
            Token::Int("5".into()),
            Token::Semicolon,
            Token::If,
            Token::LParen,
            Token::Int("5".into()),
            Token::LT,
            Token::Int("10".into()),
            Token::RParen,
            Token::LBrace,
            Token::Return,
            Token::True,
            Token::Semicolon,
            Token::RBrace,
            Token::Else,
            Token::LBrace,
            Token::Return,
            Token::False,
            Token::Semicolon,
            Token::RBrace,
            Token::Int("10".into()),
            Token::Eq,
            Token::Int("10".into()),
            Token::Semicolon,
            Token::Int("10".into()),
            Token::NotEq,
            Token::Int("9".into()),
            Token::Semicolon,
            Token::Eof,
        ];

        let mut lexer = Lexer::init(input.into());
        for token in expected_tokens {
            let token_from_lexer = lexer.next_token();
            println!(
                "token_from_lexer: {:?}, token_from_expected: {:?}",
                token_from_lexer, token
            );
            debug_assert_eq!(token_from_lexer, token);
        }
    }
}
